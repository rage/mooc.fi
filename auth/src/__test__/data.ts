import { NextFunction, Request, Response } from "express"

export const testProfile = {
  edu_person_affiliation: "",
  edu_person_principal_name: "test",
  schac_home_organization: "helsinki.fi",
  organizational_unit: "organizational_unit",
  schac_personal_unique_code:
    "urn:schac:personalUniqueCode:int:studentID:helsinki.fi:121345678",
  display_name: "test",
  given_name: "test",
  mail: "test@mail.com",
}

export const req = {
  body: {},
  query: {
    RelayState: JSON.stringify({
      language: "fi",
      provider: "hy",
      action: "sign-in",
    }),
  },
  params: {},
  login: jest.fn().mockImplementation((_: any, fn: Function) => fn(null)),
  logout: jest.fn(),
} as unknown as Request
export const res = {
  redirect: jest.fn().mockReturnThis(),
  locals: {},
  setMOOCCookies: jest.fn().mockReturnThis(),
} as unknown as Response
export const next = jest.fn() as NextFunction

export const okTokenResponse = {
  data: {
    status: 200,
    success: true,
    tmc_token: "tmc_token",
    access_token: "access_token",
    admin: false,
  },
}
