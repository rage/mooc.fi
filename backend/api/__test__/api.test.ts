import { fakeTMCCurrent, getTestContext } from "../../tests/__helpers"
import {
  normalUserDetails,
  adminUserDetails,
  thirdUserDetails,
} from "../../tests/data"
import { seed } from "../../tests/data/seed"
import axios, { Method } from "axios"
import { omit, orderBy } from "lodash"

const ctx = getTestContext()

describe("API", () => {
  interface RequestParams {
    data?: any
    headers?: any
    params?: Record<string, any>
  }
  const request = (method: Method) => (
    route: string = "",
    defaultHeaders: any,
  ) => async ({
    data = null,
    headers = defaultHeaders,
    params = {},
  }: RequestParams) =>
    await axios({
      method,
      url: `http://localhost:${ctx.port}${route}`,
      data,
      headers,
      params,
    })

  const get = (route: string = "", defaultHeaders: any) =>
    request("GET")(route, defaultHeaders)
  const post = (route: string = "", defaultHeaders: any) =>
    request("POST")(route, defaultHeaders)

  describe("/register-completions", () => {
    const defaultHeaders = {
      Authorization: "Basic kissa",
    }
    const postCompletions = post("/api/register-completions", defaultHeaders)

    beforeEach(async () => {
      await seed(ctx.prisma)
    })

    it("errors on wrong authorization", async () => {
      return postCompletions({
        data: { foo: 1 },
        headers: { Authorization: "foo" },
      })
        .then(() => fail())
        .catch(({ response }) => {
          expect(response.status).toBe(401)
        })
    })

    it("errors on non-existent secret", async () => {
      return postCompletions({
        data: { foo: 1 },
        headers: { Authorization: "Basic koira" },
      })
        .then(() => fail())
        .catch(({ response }) => {
          expect(response.status).toBe(401)
        })
    })

    it("errors on no completions", async () => {
      return postCompletions({ data: { foo: 1 } })
        .then(() => fail())
        .catch(({ response }) => {
          expect(response.status).toBe(400)
        })
    })

    it("errors on malformed completion", async () => {
      return postCompletions({
        data: {
          completions: [
            {
              foo: 1,
            },
          ],
        },
      })
        .then(() => fail())
        .catch(({ response }) => {
          expect(response.status).toBe(403)
        })
    })

    it("creates registered completions", async () => {
      const res = await postCompletions({
        data: {
          completions: [
            {
              completion_id: "30000000-0000-0000-0000-000000000102",
              student_number: "12345",
            },
            {
              completion_id: "30000000-0000-0000-0000-000000000103",
              student_number: "12345",
            },
          ],
        },
      })

      expect(res.status).toBe(200)

      const addedCompletions = await ctx.prisma.completionRegistered.findMany({
        where: {
          user_id: "20000000000000000000000000000102",
        },
      })

      expect(orderBy(addedCompletions, "completion_id")).toMatchSnapshot([
        {
          id: expect.any(String),
          created_at: expect.any(Date),
          updated_at: expect.any(Date),
        },
        {
          id: expect.any(String),
          created_at: expect.any(Date),
          updated_at: expect.any(Date),
        },
      ])
    })
  })

  describe("/user-course-settings", () => {
    const tmc = fakeTMCCurrent({
      "Bearer normal": [200, normalUserDetails],
      "Bearer admin": [200, adminUserDetails],
      "Bearer third": [200, thirdUserDetails],
    })

    describe("GET", () => {
      const getSettings = (slug: string) =>
        get(`/api/user-course-settings/${slug}`, {})

      beforeAll(() => tmc.setup())
      afterAll(() => tmc.teardown())

      beforeEach(async () => {
        await seed(ctx.prisma)
      })

      it("errors without slug", async () => {
        return getSettings("")({})
          .then(() => fail())
          .catch(({ response }) => {
            expect(response.status).toBe(400)
          })
      })

      it("errors without auth", async () => {
        return getSettings("course1")({})
          .then(() => fail())
          .catch(({ response }) => {
            expect(response.status).toBe(401)
          })
      })

      it("returns null with user with no settings", async () => {
        return getSettings("course1")({
          headers: { Authorization: "Bearer third" },
        }).then((res) => {
          expect(res.data).toBeNull()
        })
      })

      it("returns null with course with no settings", async () => {
        return getSettings("handler")({
          headers: { Authorization: "Bearer normal" },
        }).then((res) => {
          expect(res.data).toBeNull()
        })
      })

      it("warns on key clashes", async () => {
        return getSettings("course2")({
          headers: { Authorization: "Bearer admin" },
        }).then(async (_) => {
          // @ts-ignore: mock
          expect(ctx.logger.warn.mock.calls.length).toBeGreaterThan(0)
          // @ts-ignore: mock
          expect(ctx.logger.warn.mock.calls[0][0]).toContain("country")
          // @ts-ignore: mock
          expect(ctx.logger.warn.mock.calls[0][0]).toContain("research")
        })
      })

      it("returns settings correctly", async () => {
        return getSettings("course1")({
          headers: { Authorization: "Bearer admin" },
        }).then(async (res) => {
          const settings = await ctx.prisma.userCourseSetting.findFirst({
            where: {
              id: "40000000-0000-0000-0000-000000000102",
            },
          })
          const expected = {
            ...omit(settings, "other"),
            ...((settings!.other as object) ?? {}),
          }

          expect(res.data).toEqual(JSON.parse(JSON.stringify(expected)))
        })
      })
    })

    describe("POST", () => {
      const postSettings = (slug: string) =>
        post(`/api/user-course-settings/${slug}`, {})

      beforeAll(() => tmc.setup())
      afterAll(() => tmc.teardown())

      beforeEach(async () => {
        await seed(ctx.prisma)
      })

      it("errors without slug", async () => {
        return postSettings("")({
          data: { foo: 1 },
        })
          .then(() => fail())
          .catch(({ response }) => {
            expect(response.status).toBe(400)
          })
      })

      it("errors without auth", async () => {
        return postSettings("course1")({
          data: { foo: 1 },
        })
          .then(() => fail())
          .catch(({ response }) => {
            expect(response.status).toBe(401)
          })
      })

      it("errors on non-existing course", async () => {
        return postSettings("foooo")({
          data: { foo: 1 },
          headers: { Authorization: "Bearer admin" },
        })
          .then(() => fail())
          .catch(({ response }) => {
            expect(response.data.message).toContain("no course")
            expect(response.status).toBe(400)
          })
      })

      it("errors without given values", async () => {
        return postSettings("course1")({
          data: {},
          headers: { Authorization: "Bearer admin" },
        })
          .then(() => fail())
          .catch(({ response }) => {
            expect(response.data.message).toContain("must provide")
            expect(response.status).toBe(400)
          })
      })

      it("creates new setting when no existing found", async () => {
        const res = await postSettings("handler")({
          data: {
            id: "bogus",
            language: "fi",
            country: "en",
            marketing: true,
            isCat: true,
            sound: "meow",
          },
          headers: { Authorization: "Bearer normal" },
        })

        expect(res.data.message).toContain("settings created")
        expect(res.status).toBe(200)

        const createdSetting = await ctx.prisma.userCourseSetting.findFirst({
          where: {
            user_id: "20000000-0000-0000-0000-000000000102",
            course_id: "00000000-0000-0000-0000-000000000666",
          },
        })

        expect(createdSetting).not.toBeNull()
        expect(createdSetting?.id).not.toEqual("bogus")
        expect(createdSetting?.language).toEqual("fi")
        expect(createdSetting?.country).toEqual("en")
        expect(createdSetting?.marketing).toEqual(true)
        expect(createdSetting?.other).toEqual({
          isCat: true,
          sound: "meow",
        })
      })

      it("updates correctly, filters unwanted fields and shoves other fields to other, updating existing ones", async () => {
        const existingSetting = await ctx.prisma.userCourseSetting.findFirst({
          where: {
            id: "40000000-0000-0000-0000-000000000102",
          },
        })

        const res = await postSettings("course1")({
          data: {
            id: "bogus",
            language: "fi",
            country: "en",
            marketing: true,
            isCat: true,
            sound: "meow",
          },
          headers: { Authorization: "Bearer admin" },
        })

        expect(res.data.message).toContain("settings updated")
        expect(res.status).toBe(200)

        const updatedSetting = await ctx.prisma.userCourseSetting.findFirst({
          where: {
            id: "40000000-0000-0000-0000-000000000102",
          },
        })

        expect(updatedSetting!.updated_at! > existingSetting!.updated_at!).toBe(
          true,
        )
        expect(updatedSetting!.language).toBe("fi")
        expect(updatedSetting!.country).toBe("en")
        expect(updatedSetting!.marketing).toBe(true)
        expect(updatedSetting!.other).toEqual({
          hasWings: true,
          isCat: true,
          sound: "meow",
        })
      })
    })

    describe("/api/completions", () => {
      const tmc = fakeTMCCurrent({
        "Bearer normal": [200, normalUserDetails],
        "Bearer admin": [200, adminUserDetails],
      })

      beforeAll(() => tmc.setup())
      afterAll(() => tmc.teardown())

      beforeEach(async () => {
        await seed(ctx.prisma)
      })

      const getCompletions = (slug: string, registered: boolean = false) =>
        get(
          `/api/completions/${slug}${
            registered ? `?registered=${registered}` : ""
          }`,
          {},
        )

      it("errors on no auth", async () => {
        return getCompletions("course1")({})
          .then(() => fail())
          .catch(({ response }) => {
            expect(response.status).toBe(401)
          })
      })

      it("errors on non-existent organization", async () => {
        return getCompletions("course1")({
          headers: { Authorization: "Basic bogus" },
        })
          .then(() => fail())
          .catch(({ response }) => {
            expect(response.status).toBe(401)
          })
      })

      it("errors on non-basic authorization", async () => {
        return getCompletions("course1")({
          headers: { Authorization: "Bearer admin" },
        })
          .then(() => fail())
          .catch(({ response }) => {
            expect(response.status).toBe(401)
          })
      })

      it("errors on non-existent course", async () => {
        return getCompletions("bogus")({
          headers: { Authorization: "Basic kissa" },
        })
          .then(() => fail())
          .catch(({ response }) => {
            expect(response.status).toBe(404)
            expect(response.data.message).toContain("Course not found")
          })
      })

      it("returns correctly on course, skipping registered completions", async () => {
        const res = await getCompletions("course1")({
          headers: { Authorization: "Basic kissa" },
        })

        const completions = orderBy(res.data, "id")

        expect(completions.map((c) => c.id)).toEqual([
          "12400000-0000-0000-0000-000000000001",
          "30000000-0000-0000-0000-000000000102",
          "30000000-0000-0000-0000-000000000104",
        ])
      })

      it("returns correctly on course alias, skipping registered completions", async () => {
        const res = await getCompletions("alias")({
          headers: { Authorization: "Basic kissa" },
        })

        const completions = orderBy(res.data, "id")

        expect(completions.map((c) => c.id)).toEqual([
          "12400000-0000-0000-0000-000000000001",
          "30000000-0000-0000-0000-000000000102",
          "30000000-0000-0000-0000-000000000104",
        ])
      })

      it("returns correctly on course when registered query parameter is set", async () => {
        const res = await getCompletions(
          "course1",
          true,
        )({
          headers: { Authorization: "Basic kissa" },
        })

        const completions = orderBy(res.data, "id")

        expect(completions.map((c) => c.id)).toEqual([
          "12400000-0000-0000-0000-000000000001",
          "30000000-0000-0000-0000-000000000102",
          "30000000-0000-0000-0000-000000000104",
          "30000000-0000-0000-0000-000000000105",
        ])
      })

      const completionInstructions = get(
        "/api/completionInstructions/course1/en",
        {},
      )

      it("returns course instructions", async () => {
        const res = await completionInstructions({})

        expect(res.status).toBe(200)
      })

      const completionTiers = get("/api/completionTiers/course1", {})

      it("returns course with multiple tiered registration links", async () => {
        const res = await completionTiers({
          headers: { Authorization: "Bearer normal" },
        })

        expect(res.status).toBe(200)
      })
    })
  })

  describe("/ab-studies", () => {
    const tmc = fakeTMCCurrent({
      "Bearer normal": [200, normalUserDetails],
      "Bearer admin": [200, adminUserDetails],
    })

    beforeAll(() => tmc.setup())
    afterAll(() => tmc.teardown())

    beforeEach(async () => {
      await seed(ctx.prisma)
    })

    const getStudy = (id?: string) => get(`/api/ab-studies/${id ?? ""}`, {})

    const okStudyId = "99000000-0000-0000-0000-000000000001"
    const okStudyId2 = "99000000-0000-0000-0000-000000000002"
    const fakeStudyId = "99000000-0000-0000-0000-000000000003"

    it("errors on no auth", async () => {
      return getStudy(okStudyId)({})
        .then(() => fail())
        .catch(({ response }) => {
          expect(response.status).toBe(401)
        })
    })

    it("errors on non-admin", async () => {
      return getStudy(okStudyId)({
        headers: {
          Authorization: "Bearer normal",
        },
      })
        .then(() => fail())
        .catch(({ response }) => {
          expect(response.status).toBe(401)
        })
    })

    it("returns empty on non-existent study", async () => {
      const response = await getStudy(fakeStudyId)({
        headers: {
          Authorization: "Bearer admin",
        },
      })

      expect(response.status).toBe(200)
      expect(response.data).toBeNull()
    })

    it("returns existing study", async () => {
      const response = await getStudy(okStudyId)({
        headers: {
          Authorization: "Bearer admin",
        },
      })

      expect(response.status).toBe(200)
      expect((response.data as any).id).toEqual(okStudyId)
    })

    it("returns all studies on no id given", async () => {
      const response = await getStudy()({
        headers: {
          Authorization: "Bearer admin",
        },
      })

      expect(response.status).toBe(200)
      expect(Array.isArray(response.data)).toBe(true)

      const ids = response.data.map((d: any) => d.id)
      expect(ids.length).toBe(2)
      expect(ids).toContain(okStudyId)
      expect(ids).toContain(okStudyId2)
    })
  })

  describe("/ab-enrollments", () => {
    const tmc = fakeTMCCurrent({
      "Bearer normal": [200, normalUserDetails],
      "Bearer admin": [200, adminUserDetails],
    })

    beforeAll(() => tmc.setup())
    afterAll(() => tmc.teardown())

    beforeEach(async () => {
      await seed(ctx.prisma)
    })

    const getEnrollment = (id: string) => get(`/api/ab-enrollments/${id}`, {})

    const okStudyId = "99000000-0000-0000-0000-000000000001"
    const okStudyId2 = "99000000-0000-0000-0000-000000000002"
    const fakeStudyId = "99000000-0000-0000-0000-000000000003"

    it("errors on no auth", async () => {
      return getEnrollment(okStudyId)({})
        .then(() => fail())
        .catch(({ response }) => {
          expect(response.status).toBe(401)
        })
    })

    it("errors on non-existent ab_study_id", async () => {
      return getEnrollment(fakeStudyId)({
        headers: {
          Authorization: "Bearer normal",
        },
      })
        .then(() => fail())
        .catch(({ response }) => {
          expect(response.status).toBe(400)
          expect(response.data.message).toContain("not found")
        })
    })

    it("returns existing enrollment", async () => {
      const countBefore = await ctx.knex.count("*").from("ab_enrollment")

      const existingEnrollmentResponse = await getEnrollment(okStudyId2)({
        headers: {
          Authorization: "Bearer admin",
        },
      })

      const countAfter = await ctx.knex.count("*").from("ab_enrollment")
      expect(countBefore).toEqual(countAfter)

      expect(existingEnrollmentResponse.status).toBe(200)
      expect(existingEnrollmentResponse.data.group).toBe(2)
    })

    it("creates enrollment and returns the created one on future calls", async () => {
      const createdEnrollmentResponse = await getEnrollment(okStudyId)({
        headers: {
          Authorization: "Bearer normal",
        },
      })

      expect(createdEnrollmentResponse.status).toBe(200)
      expect(createdEnrollmentResponse.data).not.toBeNull()

      for (let i = 0; i < 10; i++) {
        const enrollmentResponse = await getEnrollment(okStudyId)({
          headers: {
            Authorization: "Bearer normal",
          },
        })

        expect(enrollmentResponse.data).toEqual(createdEnrollmentResponse.data)
      }
    })
  })
})
