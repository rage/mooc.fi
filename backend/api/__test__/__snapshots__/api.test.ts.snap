// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API /progressv2 returns correct data on course id, include deleted false 1`] = `
Object {
  "data": Object {
    "completion": Object {
      "certificate_id": null,
      "completion_date": "2000-01-01T08:00:00.000Z",
      "completion_language": null,
      "completion_registration_attempt_date": "2022-01-01T08:00:00.000Z",
      "course_id": "00000000-0000-0000-0000-000000000001",
      "created_at": "2000-01-01T08:00:00.000Z",
      "eligible_for_ects": true,
      "email": "e@mail.com",
      "grade": null,
      "id": "30000000-0000-0000-0000-000000000103",
      "student_number": null,
      "tier": null,
      "updated_at": "2000-01-01T08:00:00.000Z",
      "user_id": "20000000-0000-0000-0000-000000000102",
      "user_upstream_id": 1,
    },
    "course_id": "00000000-0000-0000-0000-000000000001",
    "exercise_completions": Object {
      "50000000-0000-0000-0000-000000000001": Object {
        "completed": true,
        "exercise_id": "50000000-0000-0000-0000-000000000001",
        "max_points": 2,
        "n_points": "6.000000000000000000000000000000",
        "part": 1,
        "quizzes_id": "customid1",
        "section": 1,
        "user_id": "20000000-0000-0000-0000-000000000102",
      },
      "50000000-0000-0000-0000-000000000002": Object {
        "completed": false,
        "exercise_id": "50000000-0000-0000-0000-000000000002",
        "max_points": 3,
        "n_points": "0.000000000000000000000000000000",
        "part": 1,
        "quizzes_id": "customid2",
        "section": 2,
        "user_id": "20000000-0000-0000-0000-000000000102",
      },
    },
    "user_id": "20000000-0000-0000-0000-000000000102",
  },
}
`;

exports[`API /progressv2 returns correct data on course id, include deleted true 1`] = `
Object {
  "data": Object {
    "completion": Object {
      "certificate_id": null,
      "completion_date": "2000-01-01T08:00:00.000Z",
      "completion_language": null,
      "completion_registration_attempt_date": "2022-01-01T08:00:00.000Z",
      "course_id": "00000000-0000-0000-0000-000000000001",
      "created_at": "2000-01-01T08:00:00.000Z",
      "eligible_for_ects": true,
      "email": "e@mail.com",
      "grade": null,
      "id": "30000000-0000-0000-0000-000000000103",
      "student_number": null,
      "tier": null,
      "updated_at": "2000-01-01T08:00:00.000Z",
      "user_id": "20000000-0000-0000-0000-000000000102",
      "user_upstream_id": 1,
    },
    "course_id": "00000000-0000-0000-0000-000000000001",
    "exercise_completions": Object {
      "50000000-0000-0000-0000-000000000001": Object {
        "completed": true,
        "exercise_id": "50000000-0000-0000-0000-000000000001",
        "max_points": 2,
        "n_points": "6.000000000000000000000000000000",
        "part": 1,
        "quizzes_id": "customid1",
        "section": 1,
        "user_id": "20000000-0000-0000-0000-000000000102",
      },
      "50000000-0000-0000-0000-000000000002": Object {
        "completed": false,
        "exercise_id": "50000000-0000-0000-0000-000000000002",
        "max_points": 3,
        "n_points": "0.000000000000000000000000000000",
        "part": 1,
        "quizzes_id": "customid2",
        "section": 2,
        "user_id": "20000000-0000-0000-0000-000000000102",
      },
      "50000000-0000-0000-0000-000000000005": Object {
        "completed": false,
        "exercise_id": "50000000-0000-0000-0000-000000000005",
        "max_points": null,
        "n_points": "1.000000000000000000000000000000",
        "part": null,
        "quizzes_id": "customid5",
        "section": null,
        "user_id": "20000000-0000-0000-0000-000000000102",
      },
    },
    "user_id": "20000000-0000-0000-0000-000000000102",
  },
}
`;

exports[`API /progressv2 returns correct data on course slug, include deleted false 1`] = `
Object {
  "data": Object {
    "completion": Object {
      "certificate_id": null,
      "completion_date": "2000-01-01T08:00:00.000Z",
      "completion_language": null,
      "completion_registration_attempt_date": "2022-01-01T08:00:00.000Z",
      "course_id": "00000000-0000-0000-0000-000000000001",
      "created_at": "2000-01-01T08:00:00.000Z",
      "eligible_for_ects": true,
      "email": "e@mail.com",
      "grade": null,
      "id": "30000000-0000-0000-0000-000000000103",
      "student_number": null,
      "tier": null,
      "updated_at": "2000-01-01T08:00:00.000Z",
      "user_id": "20000000-0000-0000-0000-000000000102",
      "user_upstream_id": 1,
    },
    "course_id": "00000000-0000-0000-0000-000000000001",
    "exercise_completions": Object {
      "50000000-0000-0000-0000-000000000001": Object {
        "completed": true,
        "exercise_id": "50000000-0000-0000-0000-000000000001",
        "max_points": 2,
        "n_points": "6.000000000000000000000000000000",
        "part": 1,
        "quizzes_id": "customid1",
        "section": 1,
        "user_id": "20000000-0000-0000-0000-000000000102",
      },
      "50000000-0000-0000-0000-000000000002": Object {
        "completed": false,
        "exercise_id": "50000000-0000-0000-0000-000000000002",
        "max_points": 3,
        "n_points": "0.000000000000000000000000000000",
        "part": 1,
        "quizzes_id": "customid2",
        "section": 2,
        "user_id": "20000000-0000-0000-0000-000000000102",
      },
    },
    "user_id": "20000000-0000-0000-0000-000000000102",
  },
}
`;

exports[`API /progressv2 returns correct data on course slug, include deleted true 1`] = `
Object {
  "data": Object {
    "completion": Object {
      "certificate_id": null,
      "completion_date": "2000-01-01T08:00:00.000Z",
      "completion_language": null,
      "completion_registration_attempt_date": "2022-01-01T08:00:00.000Z",
      "course_id": "00000000-0000-0000-0000-000000000001",
      "created_at": "2000-01-01T08:00:00.000Z",
      "eligible_for_ects": true,
      "email": "e@mail.com",
      "grade": null,
      "id": "30000000-0000-0000-0000-000000000103",
      "student_number": null,
      "tier": null,
      "updated_at": "2000-01-01T08:00:00.000Z",
      "user_id": "20000000-0000-0000-0000-000000000102",
      "user_upstream_id": 1,
    },
    "course_id": "00000000-0000-0000-0000-000000000001",
    "exercise_completions": Object {
      "50000000-0000-0000-0000-000000000001": Object {
        "completed": true,
        "exercise_id": "50000000-0000-0000-0000-000000000001",
        "max_points": 2,
        "n_points": "6.000000000000000000000000000000",
        "part": 1,
        "quizzes_id": "customid1",
        "section": 1,
        "user_id": "20000000-0000-0000-0000-000000000102",
      },
      "50000000-0000-0000-0000-000000000002": Object {
        "completed": false,
        "exercise_id": "50000000-0000-0000-0000-000000000002",
        "max_points": 3,
        "n_points": "0.000000000000000000000000000000",
        "part": 1,
        "quizzes_id": "customid2",
        "section": 2,
        "user_id": "20000000-0000-0000-0000-000000000102",
      },
      "50000000-0000-0000-0000-000000000005": Object {
        "completed": false,
        "exercise_id": "50000000-0000-0000-0000-000000000005",
        "max_points": null,
        "n_points": "1.000000000000000000000000000000",
        "part": null,
        "quizzes_id": "customid5",
        "section": null,
        "user_id": "20000000-0000-0000-0000-000000000102",
      },
    },
    "user_id": "20000000-0000-0000-0000-000000000102",
  },
}
`;

exports[`API /register-completions creates registered completions 1`] = `
Array [
  Object {
    "completion_id": "30000000-0000-0000-0000-000000000102",
    "course_id": "00000000-0000-0000-0000-000000000002",
    "created_at": Any<Date>,
    "id": Any<String>,
    "organization_id": "10000000-0000-0000-0000-000000000102",
    "real_student_number": "12345",
    "registration_date": null,
    "updated_at": Any<Date>,
    "user_id": "20000000-0000-0000-0000-000000000102",
  },
  Object {
    "completion_id": "30000000-0000-0000-0000-000000000103",
    "course_id": "00000000-0000-0000-0000-000000000001",
    "created_at": Any<Date>,
    "id": Any<String>,
    "organization_id": "10000000-0000-0000-0000-000000000102",
    "real_student_number": "12345",
    "registration_date": null,
    "updated_at": Any<Date>,
    "user_id": "20000000-0000-0000-0000-000000000102",
  },
]
`;

exports[`API /stored-data get returns stored data from owned course 1`] = `
Array [
  Object {
    "completions": Array [
      Object {
        "certificate_id": null,
        "completion_date": "2000-01-01T08:00:00.000Z",
        "completion_language": null,
        "completion_registration_attempt_date": "2022-01-01T08:00:00.000Z",
        "course_id": "00000000-0000-0000-0000-000000000001",
        "created_at": "2000-01-01T08:00:00.000Z",
        "eligible_for_ects": true,
        "email": "e@mail.com",
        "grade": null,
        "id": "30000000-0000-0000-0000-000000000103",
        "student_number": null,
        "tier": null,
        "updated_at": "2000-01-01T08:00:00.000Z",
        "user_id": "20000000-0000-0000-0000-000000000102",
        "user_upstream_id": 1,
      },
    ],
    "storedData": Object {
      "course_id": "00000000-0000-0000-0000-000000000001",
      "created_at": "1900-01-01T08:00:00.000Z",
      "data": "user1_foo",
      "updated_at": "1900-01-01T08:00:00.000Z",
      "user_id": "20000000-0000-0000-0000-000000000102",
    },
    "user": Object {
      "administrator": false,
      "created_at": "1900-01-01T08:00:00.000Z",
      "email": "e@mail.com",
      "first_name": null,
      "id": "20000000-0000-0000-0000-000000000102",
      "last_name": null,
      "real_student_number": null,
      "research_consent": null,
      "student_number": null,
      "updated_at": "1900-01-01T08:00:00.000Z",
      "upstream_id": 1,
      "username": "existing_user",
    },
  },
]
`;
